### ISSUE TITLE
Bulk Mutation or Dgraphloader got Missing Log Entry when Dgraph restarting

### ISSUE THREAD

**Issue by samaita** [dgraph-io/dgraph#1540]
Hi,

I am trying to do bulk mutation using Dgraph Client with total RDF around 15 millions using 0.82 Dgraph, and currently i am trying to do bulk mutation with total RDF around 50 millions using `dgraphloader` But this panic prevent me from completing the mutation. This panic appearing after the Dgraph restarting.
```
Dgraph version   : v0.8.2
Commit SHA-1     : 1063dee9
Commit timestamp : 2017-09-20 14:47:32 +1000
Branch           : HEAD

2017/09/26 04:25:35 draft.go:252: Node with GroupID: 0, ID: 1
2017/09/26 04:25:35 gRPC server started.  Listening on port 9090
2017/09/26 04:25:35 HTTP server started.  Listening on port 8080
2017/09/26 04:25:35 worker.go:112: Worker listening at address: [::]:12345
2017/09/26 04:25:35 draft.go:252: Node with GroupID: 1, ID: 1
2017/09/26 04:25:35 draft.go:1086: Found hardstate: {Term:3 Vote:1 Commit:10 XXX_unrecognized:[]}
2017/09/26 04:25:35 draft.go:1069: Found Snapshot: {Data:[9 1 0 0 0 0 0 0 0 16 1 26 15 108 111 99 97 108 104 111 115 116 58 49 50 51 52 53] Metadata:{ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:66407 Term:7 XXX_unrecognized:[]} XXX_unrecognized:[]}
2017/09/26 04:25:35 draft.go:1098: Group 0 found 5 entries
raft2017/09/26 04:25:35 missing log entry [last: 0, append at: 6]
2017/09/26 04:25:35 draft.go:1086: Found hardstate: {Term:7 Vote:1 Commit:67937 XXX_unrecognized:[]}
panic: missing log entry [last: 0, append at: 6]

goroutine 3346 [running]:
log.(*Logger).Panicf(0xc4201cc230, 0xdb9a11, 0x2b, 0xc4de3a0000, 0x2, 0x2)
	/home/travis/.gimme/versions/go1.9.linux.amd64/src/log/log.go:223 +0xdb
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*DefaultLogger).Panicf(0x14b1410, 0xdb9a11, 0x2b, 0xc4de3a0000, 0x2, 0x2)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/logger.go:121 +0x60
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*MemoryStorage).Append(0xc44cd182a0, 0xc42022e240, 0x5, 0x8, 0x0, 0x0)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/storage.go:267 +0x641
github.com/dgraph-io/dgraph/worker.(*node).initFromWal(0xc4201d6280, 0xc4dd958540, 0x14bfb80, 0x14bfb80, 0x14bfb90)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/draft.go:1102 +0x50a
github.com/dgraph-io/dgraph/worker.(*node).InitAndStartNode(0xc4201d6280, 0xc4dd958540)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/draft.go:1108 +0x4f
github.com/dgraph-io/dgraph/worker.StartRaftNodes.func2(0xc4dd957610, 0xc4201d6280)
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:171 +0x61
created by github.com/dgraph-io/dgraph/worker.StartRaftNodes
	/home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:169 +0x3a8
2017/09/26 04:25:35 main.go:166: Loading configuration from file: /data/dgraph/config.yaml
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["debugmode" = false]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["bindall" = true]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["gentlecommit" = 0.33]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["idx" = 1]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["groups" = 0,1]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["port" = 8080]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["grpc_port" = 9090]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["workerport" = 12345]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["export" = export]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["memory_mb" = 4096]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["trace" = 0.33]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["p" = p]
2017/09/26 04:25:35 init.go:73: Picked flag from config: ["w" = w]
```
I did found the similar panic listed as an issue and closed here: https://github.com/dgraph-io/dgraph/issues/11521.

Thanks in advance!

**Comment by janardhan1993** [comment]
@samaita Can you please share your w directory. I was trying to reproduce it but couldn't reproduce.

**Comment by samaita** [comment]
Took me a while to reproduce it, since the `w` directory was deleted. I was redo use dgraphloader to load 250 Millions rdf, and stopped at ~97 Millions. This is the `w` directory: https://www.dropbox.com/s/f1winqdb54rk9w3/w.tar.gz?dl=0 And this is the error log. Hope it will help. Thanks
```
Dgraph version   : v0.8.2
Commit SHA-1     : 4269f91f
Commit timestamp : 2017-09-27 14:49:31 +1000
Branch           : HEAD

2017/10/09 23:35:44 draft.go:252: Node with GroupID: 0, ID: 1
2017/10/09 23:35:44 worker.go:112: Worker listening at address: [::]:12344
2017/10/09 23:35:44 gRPC server started.  Listening on port 9090
2017/10/09 23:35:44 HTTP server started.  Listening on port 8080
2017/10/09 23:35:44 draft.go:252: Node with GroupID: 1, ID: 1
2017/10/09 23:35:44 draft.go:1086: Found hardstate: {Term:2 Vote:1 Commit:6 XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1069: Found Snapshot: {Data:[9 1 0 0 0 0 0 0 0 16 1 26 15 108 111 99 97 108 104 111 115 116 58 49 50 51 52 52] Metadata:{ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:942592 Term:2 XXX_unrecognized:[]} 
XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1086: Found hardstate: {Term:2 Vote:1 Commit:972081 XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1098: Group 1 found 12728 entries
2017/10/09 23:35:44 draft.go:1112: Restarting node for group: 1
2017/10/09 23:35:44 logger.go:121: 1 state.commit 972081 is out of range [942592, 955320]
panic: 1 state.commit 972081 is out of range [942592, 955320]

goroutine 6355 [running]:
log.(*Logger).Panicf(0xc42017c000, 0xdb9146, 0x2b, 0xc459a2a040, 0x4, 0x4)
        /home/travis/.gimme/versions/go1.9.linux.amd64/src/log/log.go:223 +0xdb
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*DefaultLogger).Panicf(0xc4430f1db0, 0xdb9146, 0x2b, 0xc459a2a040, 0x4, 0x4)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/logger.go:121 +0x60
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*raft).loadState(0xc459a26000, 0x2, 0x1, 0xed531, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/raft.go:1200 +0x1c7
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.newRaft(0xc43b212100, 0xc43b215e28)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/raft.go:303 +0xc39
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.RestartNode(0xc43b212100, 0x1, 0xc400000000)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/node.go:219 +0x43
github.com/dgraph-io/dgraph/worker.(*node).InitAndStartNode(0xc43b1943c0, 0xc4430f1bb0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/draft.go:1117 +0x148
github.com/dgraph-io/dgraph/worker.StartRaftNodes.func2(0xc4430fcc40, 0xc43b1943c0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:171 +0x61
created by github.com/dgraph-io/dgraph/worker.StartRaftNodes
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:169 +0x3a8
2017/10/09 23:35:44 main.go:166: Loading configuration from file: /data/dgraph/config.yaml
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["bindall" = true]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["export" = export]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["workerport" = 12344]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["w" = w]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["debugmode" = false]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["posting_tables" = fileio]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["memory_mb" = 2048]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["trace" = 0.33]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["grpc_port" = 9090]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["p" = p]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["gentlecommit" = 0.33]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["groups" = 0,1]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["idx" = 1]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["port" = 8080]
```

**Comment by samaita** [comment]
Took me a while to reproduce it, since the `w` directory was deleted. I was redo use dgraphloader to load 250 Millions rdf, and stopped at ~97 Millions. This is the `w` directory: https://www.dropbox.com/s/f1winqdb54rk9w3/w.tar.gz?dl=0 And this is the error log. Hope it will help. Thanks
```
Dgraph version   : v0.8.2
Commit SHA-1     : 4269f91f
Commit timestamp : 2017-09-27 14:49:31 +1000
Branch           : HEAD

2017/10/09 23:35:44 draft.go:252: Node with GroupID: 0, ID: 1
2017/10/09 23:35:44 worker.go:112: Worker listening at address: [::]:12344
2017/10/09 23:35:44 gRPC server started.  Listening on port 9090
2017/10/09 23:35:44 HTTP server started.  Listening on port 8080
2017/10/09 23:35:44 draft.go:252: Node with GroupID: 1, ID: 1
2017/10/09 23:35:44 draft.go:1086: Found hardstate: {Term:2 Vote:1 Commit:6 XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1069: Found Snapshot: {Data:[9 1 0 0 0 0 0 0 0 16 1 26 15 108 111 99 97 108 104 111 115 116 58 49 50 51 52 52] Metadata:{ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:942592 Term:2 XXX_unrecognized:[]} 
XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1086: Found hardstate: {Term:2 Vote:1 Commit:972081 XXX_unrecognized:[]}
2017/10/09 23:35:44 draft.go:1098: Group 1 found 12728 entries
2017/10/09 23:35:44 draft.go:1112: Restarting node for group: 1
2017/10/09 23:35:44 logger.go:121: 1 state.commit 972081 is out of range [942592, 955320]
panic: 1 state.commit 972081 is out of range [942592, 955320]

goroutine 6355 [running]:
log.(*Logger).Panicf(0xc42017c000, 0xdb9146, 0x2b, 0xc459a2a040, 0x4, 0x4)
        /home/travis/.gimme/versions/go1.9.linux.amd64/src/log/log.go:223 +0xdb
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*DefaultLogger).Panicf(0xc4430f1db0, 0xdb9146, 0x2b, 0xc459a2a040, 0x4, 0x4)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/logger.go:121 +0x60
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.(*raft).loadState(0xc459a26000, 0x2, 0x1, 0xed531, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/raft.go:1200 +0x1c7
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.newRaft(0xc43b212100, 0xc43b215e28)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/raft.go:303 +0xc39
github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft.RestartNode(0xc43b212100, 0x1, 0xc400000000)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/vendor/github.com/coreos/etcd/raft/node.go:219 +0x43
github.com/dgraph-io/dgraph/worker.(*node).InitAndStartNode(0xc43b1943c0, 0xc4430f1bb0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/draft.go:1117 +0x148
github.com/dgraph-io/dgraph/worker.StartRaftNodes.func2(0xc4430fcc40, 0xc43b1943c0)
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:171 +0x61
created by github.com/dgraph-io/dgraph/worker.StartRaftNodes
        /home/travis/gopath/src/github.com/dgraph-io/dgraph/worker/groups.go:169 +0x3a8
2017/10/09 23:35:44 main.go:166: Loading configuration from file: /data/dgraph/config.yaml
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["bindall" = true]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["export" = export]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["workerport" = 12344]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["w" = w]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["debugmode" = false]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["posting_tables" = fileio]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["memory_mb" = 2048]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["trace" = 0.33]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["grpc_port" = 9090]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["p" = p]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["gentlecommit" = 0.33]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["groups" = 0,1]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["idx" = 1]
2017/10/09 23:35:44 init.go:73: Picked flag from config: ["port" = 8080]
```

**Comment by pawanrawal** [comment]
Thanks for sharing this @samaita 

Could you elaborate on how dgraphloader stopped at 97 Million? Did Dgraph get killed?

**Comment by samaita** [comment]
Thanks for the quick response @pawanrawal.

It was stopped 97 Million at due to OOM. But then Dgraph keep on failed to restart due to panic.
I use this to load: `dgraphloader -d 127.0.0.1:9090 -r export/complete.rdf.gz -s export/schema.txt -m 5000`

**Comment by janardhan1993** [comment]
The restart issue is fixed now in v0.8.3 and v0.8.2. Please get latest binaries and try again. 

We will investigate OOM issue on 8GB machine.

**Comment by janardhan1993** [comment]
The issue was due to gc of file after head pointer. Fixed now.
